### PR TITLE
feat: more flexible curve fit

### LINKED ISSUE #3294: Look into Xarray curvefit wrapper
See https://github.com/orgs/scipp/discussions/3249. Xarray's `curvefit` wrapper seems quite powerful and useful. It may be an alternative to wrapping SciPy `curve_fit` directly.

### PR DESCRIPTION
**Pull request by jokasimr**
Fixes #3294 .

Implements a utility for mapping over `sc.DataArray`s and implements the broadcasted 1D `curve_fit` in terms of it.

This is a draft and currently the benchmark takes about 5x longer for scipp than the same function in xarray.
Here's the script I'm using to compare performance:

```python
import numpy as np
import scipp as sc
import scipy
from scipp.scipy.optimize import curve_fit, apply_ufunc
N = 30
x = sc.linspace('x', 0, 1, N)
y = sc.linspace('y', 0, 1, N)
t = sc.linspace('t', 0, 1, 110)
a, b = 2, -0.1
data = 0 * x + 0 * y + a * sc.exp(b * t)

da = sc.DataArray(
    data,
    coords={'x': x, 'y': y, 't': t}
)

import xarray

xda = sc.compat.to_xarray(da)

# Time the statement below to measure the time it takes to run the fit in scipp
curve_fit(lambda t, *, a, b: a * sc.exp(b * t), da, p0={'a': 1, 'b': 1})
# on my workstation (with scipp release build) 1.0s

# Time the statement below to measure the time it takes to run the fit in xarray
xda.curvefit(('t', ), lambda t, a, b: a * np.exp(b * t), p0=(1, 1))
# on my workstation 190ms
```

(Both run on a single core.)

I think there are some changes that can be done to bring us within 2x of xarray without too much effort.
I also think the two cases are not entirely comparable in terms of what work they do and that also makes up for some of the difference.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
What about #3294, i.e., making an Xarray wrapper? From what I can tell, `xarray.curvefit` can do more than just broadcast.

**Comment by jokasimr** [comment]
> What about #3294, i.e., making an Xarray wrapper? From what I can tell, `xarray.curvefit` can do more than just broadcast.

Yes that's a good option. I haven't looked into it yet.

**Comment by jokasimr** [comment]
Here is an example were a fit is made using 2 out of 3 coordinates in the data.

Screenshot from 2023-11-02 09-42-11

**Comment by jokasimr** [comment]
About performance:

A "broadcasted 1D fit" over two dimensions, many small fits, is quite slow compared to xarray curvefit (using Numpy functions to implement the fit function).

When doing a "broadcasted 2D fit" over one dimension, resulting in larger fits, the relative performance difference is lower and overall the runtime is reduced by a lot.

This indicates that what @jl-wynen said is likely the issue here. When the fit function is implemented using scipp operations the overhead is relatively large for small arrays.

File: src/scipp/scipy/optimize/__init__.py [hunk 1]
```diff
@@ -24,7 +25,20 @@ def _as_scalar(obj, unit):
     return scalar(value=obj, unit=unit)
 
 
-def _wrap_func(f, p_names, p_units):
+def _wrap_func2(f, p_names, p_units, x_units):
+    def func(x, *args):
+        if len(x.shape) == 1:
+            x = x.reshape(1, -1)
+        xs = [
+            array(dims=('row',), unit=unit, values=xi) for unit, xi in zip(x_units, x)
```
Comment (SimonHeybrock) [inline src/scipp/scipy/optimize/__init__.py:33]: This looks inefficient, why do we need to make a new variable (and thus copy/allocate memory) every time?

Comment (jokasimr) [reply src/scipp/scipy/optimize/__init__.py:33]: Hm, we could probably create the scipp arrays once outside of the fit function and only replace the `.values` here instead.

Comment (jokasimr) [reply src/scipp/scipy/optimize/__init__.py]: Not recreating the arrays did make it about 20% faster.
But it seems most of the overhead is elsewhere.

Did some profiling:

## For 1D fit for all combinations of 2 coordinates (e.g. many small fits)
**Total runtime (scipp):** 1.58 s
**Total runtime (xarray):** 189 ms
**Time spent in fit function (f) (scipp):** 750ms (50%)
**Time spent in fit function (f) (xarray):** 47ms (25%)
**Time spent in `sc.compat.to_xarray`: <1ms
**Time spent in `curve_fit` outside of the call to `xarray.Dataset.curvefit`: <1%.

Some conslusions:
    * That the time spent in the fit function is larger for scipp is not surprising. It has more overhead than the pure Numpy fit function, and the arrays we are dealing with are small. I assume this can not be improved much?
    * The difference between xarray and scipp outside of the fit functions is more difficult to explain.
    * As long as the fit function consumes 50% of the runtime any improvements outside of it will have marginal impact on total runtime.

**Comment by jokasimr** [comment]
Unfortunately we will not be able to use the variances of the data as uncertainties (`sigma` parameter) in `scipy.optimize.curve_fit`. Xarray does not have uncertainties on their arrays and don't take them into account in the fit.
We can pass `sigma` as a kwarg that xarray forwards to `scipy.optimize.curve_fit`, but that doesn't help us here because xarray might call `curve_fit` several times with different slices of the data and we need to pass different `sigma` to every such call.

I see two solutions two the problem:
 * Either we don't use variances in the fit. That would be unfortunate because it is one of the strong selling points for doing the fit in scipp instead of converting the data to xarray and doing the fit there directly.
  * Or we write our own `scipy` wrapper for fitting N-D scipp `DataArrays` without going through xarray. This is the approach I think we should take because I think fitting data to a model is a common use case. But it would require more work.

**Comment by jokasimr** [comment]
TODO
 * Do we need to make breaking changes to the old `curve_fit` api?
   * Breaking changes right now:
     * First positional argument is `coords`, a list of the coordinates that participate in the fit. We could maybe allow omitting this argument if the data is 1D, then the behavior would be the same as the old curve fit in that case?
     *  `p0` argument renamed `params`, this can be easily changed back to `p0`.
 * Rewrite documentation and add examples of fitting in several dimensions.
 * Document the `unsafe_numpy_func` escape hatch.
 * Add type hints to arguments (basically same as in the old version).
 * Add tests fitting in several dimensions.

**Comment by SimonHeybrock** [comment]
> TODO
> 
>     * Do we need to make breaking changes to the old `curve_fit` api?

Why would we? This seems unrelated (if we go with the Xarray wrapper).


> I see two solutions two the problem:
> - Either we don't use variances in the fit. That would be unfortunate because it is one of the strong selling points for doing the fit in scipp instead of converting the data to xarray and doing the fit there directly.
> - Or we write our own scipy wrapper for fitting N-D scipp DataArrays without going through xarray. This is the approach I think we should take because I think fitting data to a model is a common use case. But it would require more work.

By now you have probably developed a better understanding what Xarray's wrapper does. Would it be hard to replicate, in you opinion?

**Comment by jokasimr** [comment]
> Why would we (change the api)? This seems unrelated

The old api does not have the `coords` argument as the first argument to `curve_fit`. But the new `curve_fit` mimics `xarray` and has this argument as the first argument to the function.
We could add it as a kw arg instead.

File: src/scipp/scipy/optimize/__init__.py [hunk 2]
```diff
@@ -106,18 +173,83 @@ def _parse_bounds(
...
+        [
+            np.broadcast_to(
+                da.coords[c].values.reshape(
+                    *(da.sizes[d] if d in da.coords[c].dims else 1 for d in da.dims)
+                ),
+                da.shape,
+            ).ravel()
+            for c in coords
+        ]
+    )
```
Comment (SimonHeybrock) [inline src/scipp/scipy/optimize/__init__.py:219]: Check if `da.flatten` (potentially followed by a `fold`) can do this more concisely.

**Review by SimonHeybrock — COMMENTED** [review]
I think we should probably not extend the `scipy.optimize.curve_fit`, which is supposed to be a wrapper with a function signature that closely mimics that of SciPy. Instead, we should consider adding this as `scipp.curve_fit` or `scipp.optimize.curve_fit`.

Before a closer review, I think it is best if you could explain the overall workings of what this does, especially the various cases, face to face.

**Comment by jokasimr** [comment]
> I think we should probably not extend the `scipy.optimize.curve_fit`, which is supposed to be a wrapper with a function signature that closely mimics that of SciPy. Instead, we should consider adding this as `scipp.curve_fit` or `scipp.optimize.curve_fit`.

Should we keep the old curve fit in `scipp.scipy.optimize.curve_fit` and add the new one at `scipp.optimize.curve_fit`?
It's good because it's backwards compatible. But it might be a bit confusing to users to have two different curve fit functions with so similar names.

**Comment by SimonHeybrock** [comment]
As mentioned before, I think we should make this new and more powerful wrapper a new function, not in the `scipy` submodule. The latter is intended for wrappers to stick very close to the SciPy API.

File: src/scipp/curve_fit.py [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+        coords=coords,
+        masks={
+            m: da.masks[m]
+            for m in da.masks
+            if dims_participating_in_fit.intersection(da.masks[m].dims)
+        },
+    )
+
+    for m in _da.masks.values():
+        _da = _da[~m]
```
Comment (SimonHeybrock) [inline src/scipp/curve_fit.py:502]: Won't this fail if a mask is not 1D? I don't think there is an easy way to fix this, just wondering if you thought about this, or added a test?

File: tests/curve_fit_test.py [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+    data.values += noise_scale * np.clip(rng.normal(size=data.values.shape), -1.5, 1.5)
+    return sc.DataArray(data, coords=da_coords)
+
+
+def array1d(*, a=1.2, b=1.3, noise_scale=0.1, size=50):
+    return array(
+        dict(x=dict(dim='xx', start=-0.1, stop=4.0, num=size)),
+        func,
+        dict(a=a, b=b),
+        noise_scale=noise_scale,
```
Comment (SimonHeybrock) [inline tests/curve_fit_test.py:57]: Using keyword args seems like a good idea here, as it is this is hard to interpret.

File: tests/curve_fit_test.py [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+    }
+    data = f(**da_coords, **params)
+    # Noise is random but avoiding unbounded values to avoid flaky tests
+    data.values += noise_scale * np.clip(rng.normal(size=data.values.shape), -1.5, 1.5)
+    return sc.DataArray(data, coords=da_coords)
+
+
+def array1d(*, a=1.2, b=1.3, noise_scale=0.1, size=50):
+    return array(
+        dict(x=dict(dim='xx', start=-0.1, stop=4.0, num=size)),
```
Comment (SimonHeybrock) [inline tests/curve_fit_test.py:54]: Why is the coord name not matching the dim?

File: tests/curve_fit_test.py [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+    "coords, reduce_dims",
+    (
+        (['x'], []),
+        (['x'], ['y']),
+        (['x'], ['t', 'y']),
+        (['x', 't'], []),
+        (['x', 't'], ['y']),
+        (['x', 't', 'y'], []),
+    ),
+)
```
Comment (SimonHeybrock) [inline tests/curve_fit_test.py:125]: :+1:

File: tests/curve_fit_test.py [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+        3: (func3d, lambda x, a, b: func3d_np(*x, a, b)),
+    }[len(coords)]
+    da = array(
+        {c: dict(start=1, stop=3, num=n) for c, n in dims.items()},
+        lambda **x: (
+            sc.broadcast(
+                f(**{c: x[c] for c in x if c in coords or c in params}),
+                sizes=dims,
+            ).copy()
+        ),
```
Comment (SimonHeybrock) [inline tests/curve_fit_test.py:139]: keyword args! It is completely unclear what this does otherwise.

What is this function? Why is there a broadcast?

**Review by SimonHeybrock — COMMENTED** [review]
Looks close to done! Some minor questions/comments:

File: src/scipp/__init__.py [hunk 8]
```diff
@@ -243,6 +243,7 @@
 
 del core
 
+from .curve_fit import curve_fit
```
[the remark this answers is not in the document]
Comment (jokasimr) [reply src/scipp/__init__.py:246]: Good question. Spontaneously I'd say Scipy because that's the bigger package and more people are likely familiar with it. On the other hand we are copying the Xarray api, so having the same name maybe communicates better that they are similar? I could go with either.

File: src/scipp/curve_fit.py [hunk 9]
```diff
@@ -0,0 +1,520 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock
+"""Sub-package for optimization such as curve fitting.
+
+This subpackage provides wrappers for a subset of functions from
+:py:mod:`scipy.optimize`.
```
[the remark this answers is not in the document]
Comment (jokasimr) [reply src/scipp/curve_fit.py:7]: Oops!

File: src/scipp/curve_fit.py [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+    ] = None,
+    reduce_dims: Sequence[str] = tuple(),
+    unsafe_numpy_f: bool = False,
+    **kwargs,
+) -> Tuple[DataGroup, DataGroup]:
+    """Use non-linear least squares to fit a function, f, to data.
+
+    .. versionadded:: 23.12.0
+
+    This is a wrapper around :py:func:`scipy.optimize.curve_fit`. See there for a
```
[the remark this answers is not in the document]
Comment (jokasimr) [reply src/scipp/curve_fit.py:257]: Yes that is a good idea. We can write that it is similar to that api. There are some minor differences however. The xarray api passes coords by position, and the first argument to the fit function is a 2D array (one row for each coord participating in the fit). In our version all coords participating in the fit are separate arguments to the fit function.

File: src/scipp/curve_fit.py (hunk 3 above)
Comment (jokasimr) [reply src/scipp/curve_fit.py:502]: Hm, I did not think about that.
One alternative is to only apply the 1d masks and then set values masked by 2D masks to `nan`. The Scipy `curve_fit` function will then ignore the missing values.

File: tests/curve_fit_test.py (hunk 5 above)
Comment (jokasimr) [reply tests/curve_fit_test.py:54]: Because in general they don't have the same name, and I want to make sure the tests pass in the general case and not only in the special case when they have the same name.

Comment (SimonHeybrock) [reply tests/curve_fit_test.py:54]: Maybe make it more explicitly different then. As it is now it looks like a typo that a future editor might "fix".

File: tests/curve_fit_test.py (hunk 7 above)
Comment (jokasimr) [reply tests/curve_fit_test.py:139]: I'll add kwargs. The `array` helper takes a `func` argument that creates the data of the array from the coords.
If the function has fewer arguments than the array has dimensions then we need to broadcast it to the right shape.

Although, I realized now that if we just use `func3d` in all cases then the result will always have the right shape anyway and it's much easier to read.

Comment (SimonHeybrock) [reply tests/curve_fit_test.py:139]: I think broadcasting might hide a category of potential bugs, so using `func3d` sounds better.

File: src/scipp/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/__init__.py:246]: @scipp/scipp-maintainers Any opinions or insights from anyone?

File: tests/curve_fit_test.py [hunk 11]
```diff
@@ -163,11 +155,6 @@ def test_compare_to_curve_fit_xarray(dims, coords, reduce_dims, p0, params, boun
             result[param_name].data,
             xresult['param', sc.scalar(param_name)].data,
         )
-        if (
-            bounds is None
-            or bounds[param_name][0] <= params[param_name] <= bounds[param_name][1]
-        ):
-            assert sc.allclose(result[param_name].data, sc.scalar(params[param_name]))
```
Comment (SimonHeybrock) [inline tests/curve_fit_test.py]: Why did you remove this now?

Comment (jokasimr) [reply tests/curve_fit_test.py]: Because after replacing `func` and `func2d` with `func3d` in the array creation the result of the fit will depend on `t` and `y` in a way that I think is difficult to predict. Therefore I only kept the comparison with xarray. We test the accuracy of the results in other tests anyway.

File: src/scipp/curve_fit.py (hunk 3 above)
Comment (SimonHeybrock) [reply src/scipp/curve_fit.py:502]: I would not change anything until we have a concrete requirement, maybe just add a test to ensure that it fails as expected and does not silently do something wrong.

Comment (jokasimr) [reply src/scipp/curve_fit.py]: Fixed it by applying the masks to the flattened dataarray instead of to the nd-array.

File: tests/curve_fit_test.py [hunk 12]
```diff
@@ -492,3 +492,38 @@ def test_can_use_non_coord_in_fit():
...
+    da = array2d(a=1.2, b=1.3, noise_scale=noise_scale)
+    mask = (
+        (da.coords['x'] > 1)
+        & (da.coords['x'] < 2)
+        & (da.coords['t'] > 0.4)
+        & (da.coords['t'] < 0.6)
+    )
+    da.data += sc.where(mask, sc.scalar(1000_000), sc.scalar(0))
+    da.masks['high'] = mask
+    res, _ = curve_fit(['x', 't'], func2d, da)
```
Comment (SimonHeybrock) [inline tests/curve_fit_test.py:523]: So why is this working and not raising?

Comment (SimonHeybrock) [reply tests/curve_fit_test.py:523]: I expected it to raise since somehwere you use `da = da[~mask]`, which should not work unless the mast is 1-D.

Comment (jokasimr) [reply tests/curve_fit_test.py:523]: Aha I see. Yes that did raise previously for this test. But I fixed that in the same commit that the `test_2d_mask` was added in.

Comment (SimonHeybrock) [reply tests/curve_fit_test.py:523]: I see, since we flatten anyway we can drop the masked entries in this manner?

Comment (jokasimr) [reply tests/curve_fit_test.py:523]: Yes that's the idea.

File: src/scipp/curve_fit.py [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+        operations. If `unsafe_numpy_f` is set to `True` then the arguments passed
+        to `f` will be Numpy arrays instead of scipp Variables.
+
+    Returns
+    -------
+    popt:
+        Optimal values for the parameters.
+    pcov:
+        The estimated covariance of popt.
+
```
Comment (SimonHeybrock) [inline src/scipp/curve_fit.py:322]: One final thing: Can you add a `See Also` section here and in `scipp.scipy.optimize.curve_fit` and reference each other?

Comment (jokasimr) [reply src/scipp/curve_fit.py:331]: Should I refer to `scipp.scipy.optimize.curve_fit` as a function we keep for backwards compatibility?

Comment (SimonHeybrock) [reply src/scipp/curve_fit.py:332]: I'd just put a link with no comment?

Comment (jokasimr) [reply src/scipp/curve_fit.py:332]: I think it is good to explain the difference between them in a short comment in the See Also. Then the user will have an easier time navigating the documentation.

File: src/scipp/scipy/optimize/__init__.py [hunk 14]
```diff
@@ -174,6 +174,11 @@ def curve_fit(
     pcov:
         The estimated covariance of popt.
 
+    See Also
+    --------
+    scipp.scipy.curve_fit:
```
Comment (SimonHeybrock) [inline src/scipp/scipy/optimize/__init__.py:179]:
```suggestion
    scipp.curve_fit:
```
?

**Review by SimonHeybrock — APPROVED** [review]
